src/Transport/Sending/MessageDispatcher.cs

```diff
@@ -66,6 +71,36 @@ public Task Dispatch(TransportOperations outgoingMessages, TransportTransaction
             return tasks.Count == 1 ? tasks[0] : Task.WhenAll(tasks);
         }
 
+        static void AssertBelowMaxMessageThresholdForTransaction(List<UnicastTransportOperation> unicastTransportOperations, List<MulticastTransportOperation> multicastTransportOperations, CommittableTransaction committableTransaction)
+        {
+            var totalNumberOfOperations = unicastTransportOperations.Count + multicastTransportOperations.Count;
+            if (committableTransaction == null || totalNumberOfOperations <= MaxMessageThresholdForTransaction)
+            {
+                return;
+            }
+
+            var numberOfTransactionalOperations = 0;
+            foreach (var transportOperation in unicastTransportOperations)
```

Comment:
Should we be worried about the double enumeration we now have? I think it could be ok because we are only doing it in a very specific niche scenario to add more safety

Reply:
I wonder why we can't directly count and assert in existing loops. We have all the information available like transaction and operation count, the dispatch type etc

Reply:
If we do that, we need to remove the dispatch from the loop. I suppose we could still dispatch the first 100 and then throw when / if we get to the limit.

Reply:
I think it is better not do any IO you are right because that is more expensive that double iteration. Let's leave it

Reply:
Raised https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/841 to do that

Reply:
Did you see https://github.com/Particular/NServiceBus.Transport.AzureServiceBus/pull/840#discussion_r1246111434 😁?

Reply:
Not until after I had done the work :)